core/config/config_helper.hpp

```diff
@@ -0,0 +1,220 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_CORE_CONFIG_CONFIG_HELPER_HPP_
+#define GKO_CORE_CONFIG_CONFIG_HELPER_HPP_
+
+
+#include <string>
+#include <type_traits>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/config/config.hpp>
+#include <ginkgo/core/config/registry.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+namespace gko {
+namespace config {
+
+
+/**
+ * LinOpFactoryType enum is to avoid forward declartion, linopfactory header,
+ * two template versions of build_from_config
+ */
+enum class LinOpFactoryType : int { Cg = 0 };
+
+
+/**
+ * It is only an intermediate step after dispatching the class base type. Each
+ * implementation needs to deal with the template selection.
+ */
+template <LinOpFactoryType flag>
+deferred_factory_parameter<gko::LinOpFactory> build_from_config(
+    const pnode& config, const registry& context,
+    type_descriptor td = type_descriptor{"void", "void"});
+
+/**
+ * This function is to update the default type setting from current config.
+ *
+ * @note It might update the unused type for the current class.
```

Comment:
what does this note mean? What is the unused type?

Reply:
For example, Cg only has the ValueType but user can still put IndexType there and it will affect the default type for its child. It may help when someone wants to create several child objects with certain IndexType.

Reply:
But it might or might not update any type, so explicitly noting that it might update the unused type seems pointless. This function is already somewhat clear in what it does. It takes a `type_descriptor` and returns a new `type_descriptor` with any type updated that is defined in the config. This note doesn't add any information of value.